client/signup/controller.js

```diff
@@ -233,6 +234,26 @@ export default {
 
 		store.set( 'signup-locale', localeFromParams );
 
+		/**
+		 * The experiment is only loaded on the onboarding flow
+		 * If user is logged out we load the experiment
+		 * If user is logged in we load the experiment only if the user has no sites
+		 * More info: pbxNRc-3xO-p2
+		 */
+		const isNewUser = ! getCurrentUserSiteCount( context.store.getState() );
+		initialContext.isSignupSurveyActive = false;
+		const isOnboardingFlow = flowName === 'onboarding';
+		if ( isOnboardingFlow && ( ! userLoggedIn || ( userLoggedIn && isNewUser ) ) ) {
+			const experiment = await loadExperimentAssignment(
+				'calypso_signup_onboarding_site_goals_survey'
+			);
+			initialContext.isSignupSurveyActive = experiment.variationName === 'treatment';
+		}
```

Comment:
For posterity, I've also confirmed the following cases:

1. When `loadExperimentAssignment` somehow stuck, the signup flow will proceed as it is.
2. When `loadExperimentAssignment` throws, the signup flow will proceed as it is.
3. Cherry-picking some other /start flows, they all work as expected.